{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Graph database builder for AtgO Ontology Viewer\n",
    "\n",
    "## Introduction\n",
    "\n",
    "This is a script to generate AtgO database (DAG and raw interactions) with Neo4j v1.9.x.\n",
    "\n",
    "This notebook is the V3.  Additional information will be added from Gene Association file, including:\n",
    "\n",
    "* Full gene name\n",
    "* Synonyms\n",
    "* SGD ID\n",
    "\n",
    "## Log\n",
    "\n",
    "* 8/22/2016 - Final version for the camera-ready paper\n",
    "* 3/2/2016 - Adding new data, such as publications.\n",
    "* 3/11/2016 - first working version.\n",
    "* 4/14/2016 - missing SGD ID problems fixed.\n",
    "* 4/25/2016 - New version for updated data files.\n",
    "\n",
    "----\n",
    "## Requirments\n",
    "* Neo4j 1.9.9\n",
    "* py2neo 2.0.x\n",
    "* Rexster 2.4.x\n",
    "\n",
    "\n",
    "## How to use\n",
    "1. Install Neo4j 1.9.9 to the local machine\n",
    "1. ```pip install py2neo==2.0.9```\n",
    "1. Run Neo4j with ```./bin/neo4j start```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(Node, 'http://localhost:7474/db/data/index/node/Vertex')\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "from py2neo import Graph\n",
    "from py2neo import Node, Relationship\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# For legacy INDEXER\n",
    "from py2neo import neo4j\n",
    "\n",
    "# Start from scratch...\n",
    "g = Graph()\n",
    "g.delete_all()\n",
    "\n",
    "# Create index object\n",
    "idx = g.legacy.get_or_create_index(neo4j.Node, \"Vertex\")\n",
    "\n",
    "print(idx)\n",
    "print(g.size)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Ontology DAG in Neo4j"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term Number</th>\n",
       "      <th>Term Name</th>\n",
       "      <th>Similarity Score</th>\n",
       "      <th>Term Size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.446</td>\n",
       "      <td>484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1.941</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>2.927</td>\n",
       "      <td>384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>3.070</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Term Number Term Name  Similarity Score  Term Size\n",
       "0            1         1             0.000        492\n",
       "1            2         2             1.446        484\n",
       "2            3         3             1.941        476\n",
       "3            5         5             2.927        384\n",
       "4            6         6             3.070        156"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "old_df_term_props = pd.read_csv('./atgo2_term_info.txt', sep=\"\\t\")\n",
    "old_df_term_props.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term Number</th>\n",
       "      <th>Similarity Score</th>\n",
       "      <th>Num Genes</th>\n",
       "      <th>Curated name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>2</td>\n",
       "      <td>1.446</td>\n",
       "      <td>489</td>\n",
       "      <td>AtgO:2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>28</td>\n",
       "      <td>1.415</td>\n",
       "      <td>22</td>\n",
       "      <td>AtgO:28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>45</td>\n",
       "      <td>1.412</td>\n",
       "      <td>8</td>\n",
       "      <td>AtgO:45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>36</td>\n",
       "      <td>1.400</td>\n",
       "      <td>9</td>\n",
       "      <td>AtgO:36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>496</td>\n",
       "      <td>AtgO:1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Term Number  Similarity Score  Num Genes Curated name\n",
       "215            2             1.446        489       AtgO:2\n",
       "216           28             1.415         22      AtgO:28\n",
       "217           45             1.412          8      AtgO:45\n",
       "218           36             1.400          9      AtgO:36\n",
       "219            1             0.000        496       AtgO:1"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read properties for AtgO terms\n",
    "# df_term_props = pd.read_csv('./atgo2_term_info.txt', sep=\"\\t\")\n",
    "\n",
    "df_term_props = pd.read_csv('./data/8_22_2016/atgo2_term_info.txt', sep=\"\\t\")\n",
    "df_term_props.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "716\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1516</th>\n",
       "      <td>1</td>\n",
       "      <td>227</td>\n",
       "      <td>default</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1517</th>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>default</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518</th>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>default</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1519</th>\n",
       "      <td>1</td>\n",
       "      <td>225</td>\n",
       "      <td>default</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>default</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      target source     type  score\n",
       "1516       1    227  default    0.0\n",
       "1517       1     29  default    0.0\n",
       "1518       1     28  default    0.0\n",
       "1519       1    225  default    0.0\n",
       "1520       1      2  default    0.0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df = pd.read_csv('./AtgO2.txt', sep=\"\\t\", names=[\"target\", \"source\", \"type\"])\n",
    "\n",
    "df_graph = pd.read_csv('./data/8_22_2016/AtgO2.txt', sep=\"\\t\", names=[\"target\", \"source\", \"type\", \"score\"])\n",
    "\n",
    "nodes = []\n",
    "\n",
    "s = df_graph[\"source\"].astype(str)\n",
    "t = df_graph[\"target\"].astype(str)\n",
    "\n",
    "all_nodes = s.append(t).unique()\n",
    "\n",
    "print(len(all_nodes))\n",
    "df_graph.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term</th>\n",
       "      <th>genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>168</td>\n",
       "      <td>CDC5|MCD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>78</td>\n",
       "      <td>TAX4|IRS4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>217</td>\n",
       "      <td>IPL1|BIR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>218</td>\n",
       "      <td>VAC8|YCK3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>215</td>\n",
       "      <td>SSH4|EAR1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   term      genes\n",
       "0   168  CDC5|MCD1\n",
       "1    78  TAX4|IRS4\n",
       "2   217  IPL1|BIR1\n",
       "3   218  VAC8|YCK3\n",
       "4   215  SSH4|EAR1"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assign genes\n",
    "#df2 = pd.read_csv('AtgO2_terms_to_genes.txt', sep=\"\\t\", names=[\"term\", \"genes\"])\n",
    "\n",
    "df2 = pd.read_csv('data/8_22_2016/AtgO2_terms_to_genes.txt', sep=\"\\t\", names=[\"term\", \"genes\"])\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TAX4', 'IRS4']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assigned_genes = {}\n",
    "\n",
    "for row in df2.itertuples():\n",
    "    genes = row[2]\n",
    "    gene_list = genes.split('|')\n",
    "    assigned_genes[str(row[1])] = gene_list\n",
    "    \n",
    "assigned_genes[\"78\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kono/anaconda/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2723: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DB</th>\n",
       "      <th>DB_Object_ID</th>\n",
       "      <th>DB_Object_Symbol</th>\n",
       "      <th>Qualifier</th>\n",
       "      <th>GO ID</th>\n",
       "      <th>DB:Reference</th>\n",
       "      <th>Evidence</th>\n",
       "      <th>With (or) From</th>\n",
       "      <th>Aspect</th>\n",
       "      <th>DB_Object_Name</th>\n",
       "      <th>DB_Object_Synonym</th>\n",
       "      <th>DB_Object_Type</th>\n",
       "      <th>taxon</th>\n",
       "      <th>Date</th>\n",
       "      <th>Assigned_by</th>\n",
       "      <th>Annotation Extension</th>\n",
       "      <th>Gene Product Form ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SGD</td>\n",
       "      <td>S000007287</td>\n",
       "      <td>15S_RRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0005763</td>\n",
       "      <td>SGD_REF:S000073641|PMID:6262728</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>Ribosomal RNA of the small mitochondrial ribos...</td>\n",
       "      <td>Q0020|14s rRNA|15S_RRNA_2</td>\n",
       "      <td>gene</td>\n",
       "      <td>taxon:559292</td>\n",
       "      <td>20150612</td>\n",
       "      <td>SGD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SGD</td>\n",
       "      <td>S000007287</td>\n",
       "      <td>15S_RRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0032543</td>\n",
       "      <td>SGD_REF:S000073641|PMID:6262728</td>\n",
       "      <td>IC</td>\n",
       "      <td>GO:0005763</td>\n",
       "      <td>P</td>\n",
       "      <td>Ribosomal RNA of the small mitochondrial ribos...</td>\n",
       "      <td>Q0020|14s rRNA|15S_RRNA_2</td>\n",
       "      <td>gene</td>\n",
       "      <td>taxon:559292</td>\n",
       "      <td>20150612</td>\n",
       "      <td>SGD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SGD</td>\n",
       "      <td>S000007287</td>\n",
       "      <td>15S_RRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0003735</td>\n",
       "      <td>SGD_REF:S000073641|PMID:6262728</td>\n",
       "      <td>IC</td>\n",
       "      <td>GO:0005763</td>\n",
       "      <td>F</td>\n",
       "      <td>Ribosomal RNA of the small mitochondrial ribos...</td>\n",
       "      <td>Q0020|14s rRNA|15S_RRNA_2</td>\n",
       "      <td>gene</td>\n",
       "      <td>taxon:559292</td>\n",
       "      <td>20150612</td>\n",
       "      <td>SGD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SGD</td>\n",
       "      <td>S000007288</td>\n",
       "      <td>21S_RRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0005762</td>\n",
       "      <td>SGD_REF:S000073372|PMID:6759872</td>\n",
       "      <td>IDA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>Mitochondrial 21S rRNA</td>\n",
       "      <td>Q0158|21S_rRNA_3|21S_rRNA_4</td>\n",
       "      <td>gene</td>\n",
       "      <td>taxon:559292</td>\n",
       "      <td>20040202</td>\n",
       "      <td>SGD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SGD</td>\n",
       "      <td>S000007288</td>\n",
       "      <td>21S_RRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0032543</td>\n",
       "      <td>SGD_REF:S000073372|PMID:6759872</td>\n",
       "      <td>IMP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>Mitochondrial 21S rRNA</td>\n",
       "      <td>Q0158|21S_rRNA_3|21S_rRNA_4</td>\n",
       "      <td>gene</td>\n",
       "      <td>taxon:559292</td>\n",
       "      <td>20100715</td>\n",
       "      <td>SGD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    DB DB_Object_ID DB_Object_Symbol Qualifier       GO ID  \\\n",
       "0  SGD   S000007287         15S_RRNA       NaN  GO:0005763   \n",
       "1  SGD   S000007287         15S_RRNA       NaN  GO:0032543   \n",
       "2  SGD   S000007287         15S_RRNA       NaN  GO:0003735   \n",
       "3  SGD   S000007288         21S_RRNA       NaN  GO:0005762   \n",
       "4  SGD   S000007288         21S_RRNA       NaN  GO:0032543   \n",
       "\n",
       "                      DB:Reference Evidence With (or) From Aspect  \\\n",
       "0  SGD_REF:S000073641|PMID:6262728      IDA            NaN      C   \n",
       "1  SGD_REF:S000073641|PMID:6262728       IC     GO:0005763      P   \n",
       "2  SGD_REF:S000073641|PMID:6262728       IC     GO:0005763      F   \n",
       "3  SGD_REF:S000073372|PMID:6759872      IDA            NaN      C   \n",
       "4  SGD_REF:S000073372|PMID:6759872      IMP            NaN      P   \n",
       "\n",
       "                                      DB_Object_Name  \\\n",
       "0  Ribosomal RNA of the small mitochondrial ribos...   \n",
       "1  Ribosomal RNA of the small mitochondrial ribos...   \n",
       "2  Ribosomal RNA of the small mitochondrial ribos...   \n",
       "3                             Mitochondrial 21S rRNA   \n",
       "4                             Mitochondrial 21S rRNA   \n",
       "\n",
       "             DB_Object_Synonym DB_Object_Type         taxon      Date  \\\n",
       "0    Q0020|14s rRNA|15S_RRNA_2           gene  taxon:559292  20150612   \n",
       "1    Q0020|14s rRNA|15S_RRNA_2           gene  taxon:559292  20150612   \n",
       "2    Q0020|14s rRNA|15S_RRNA_2           gene  taxon:559292  20150612   \n",
       "3  Q0158|21S_rRNA_3|21S_rRNA_4           gene  taxon:559292  20040202   \n",
       "4  Q0158|21S_rRNA_3|21S_rRNA_4           gene  taxon:559292  20100715   \n",
       "\n",
       "  Assigned_by  Annotation Extension  Gene Product Form ID  \n",
       "0         SGD                   NaN                   NaN  \n",
       "1         SGD                   NaN                   NaN  \n",
       "2         SGD                   NaN                   NaN  \n",
       "3         SGD                   NaN                   NaN  \n",
       "4         SGD                   NaN                   NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract gene info from GA file\n",
    "HEADER = [\"DB\", \"DB_Object_ID\", \"DB_Object_Symbol\", \"Qualifier\", \"GO ID\", \"DB:Reference\", \"Evidence\", \"With (or) From\",\n",
    "    \"Aspect\", \"DB_Object_Name\", \"DB_Object_Synonym\", \"DB_Object_Type\", \"taxon\", \"Date\", \"Assigned_by\", \n",
    "          \"Annotation Extension\", \"Gene Product Form ID\"]\n",
    "df_ga = pd.read_csv('./gene_association.sgd', sep=\"\\t\", comment=\"!\", header=None, names=HEADER);\n",
    "\n",
    "df_ga.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gene_details = {}\n",
    "\n",
    "for row in df_ga.itertuples():\n",
    "    details = {}\n",
    "    sgd_id = row[2]\n",
    "    key = row[3]\n",
    "    full_name = row[10]\n",
    "    synonym = str(row[11])\n",
    "    details[\"sgd\"] = str(sgd_id)\n",
    "    details[\"name\"] = str(full_name)\n",
    "    syn_list = synonym.split('|')\n",
    "    details[\"syn_list\"] = syn_list\n",
    "    \n",
    "    gene_details[key] = details\n",
    "    # Use aliases, too\n",
    "    for syn in syn_list:\n",
    "         gene_details[syn] = details\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'syn_list': ['YFR011C', 'AIM13', 'MCS19'], 'name': 'Component of the MICOS complex', 'sgd': 'S000001907'}\n",
      "{'syn_list': ['YBR262C', 'AIM5', 'FMP51', 'MCS12'], 'name': 'Component of the MICOS complex', 'sgd': 'S000000466'}\n",
      "{'syn_list': ['YNL100W', 'AIM37', 'MCS27'], 'name': 'Component of the MICOS complex', 'sgd': 'S000005044'}\n",
      "{'syn_list': ['YKR016W', 'AIM28', 'FCJ1', 'FMP13'], 'name': 'Component of the MICOS complex', 'sgd': 'S000001724'}\n",
      "{'syn_list': ['YDR108W', 'GSG1', 'MUM1'], 'name': 'Component of transport protein particle (TRAPP) complex III', 'sgd': 'S000002515'}\n",
      "{'syn_list': ['YDR022C', 'CIS1'], 'name': 'Autophagy-specific protein required for autophagosome formation', 'sgd': 'S000002429'}\n",
      "{'syn_list': ['YDR022C', 'CIS1'], 'name': 'Autophagy-specific protein required for autophagosome formation', 'sgd': 'S000002429'}\n"
     ]
    }
   ],
   "source": [
    "# Check gene list is complete\n",
    "print(gene_details['AIM13'])\n",
    "print(gene_details['AIM5'])\n",
    "print(gene_details['AIM37'])\n",
    "print(gene_details['FCJ1'])\n",
    "print(gene_details['GSG1'])\n",
    "\n",
    "# Swap special case for CIS1\n",
    "original_entry = gene_details['ATG31']\n",
    "gene_details['CIS1'] = original_entry\n",
    "\n",
    "print(gene_details['ATG31'])\n",
    "print(gene_details['CIS1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TOR complex catalytic subunits\n"
     ]
    }
   ],
   "source": [
    "# Test find term info\n",
    "ti = df_term_props.loc[df_term_props[\"Term Number\"] == 243]\n",
    "print(ti[\"Curated name\"].values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def is_int(str_val):\n",
    "    try: \n",
    "        int(str_val)\n",
    "        return True\n",
    "    except ValueError:\n",
    "        return False\n",
    "\n",
    "def add_node(node_name, assigned_genes, index, df_t):\n",
    "    \n",
    "    node_props = {\n",
    "        \"name\": str(node_name)\n",
    "    }\n",
    "    \n",
    "    if is_int(node_name):\n",
    "        # This is an ontology term\n",
    "        tid = int(node_name)\n",
    "        term_info = df_t.loc[df_t[\"Term Number\"] == tid]\n",
    "    \n",
    "        if term_info.empty == False:\n",
    "            node_props[\"term_name\"] = str(term_info[\"Curated name\"].values[0])\n",
    "            node_props[\"similarity_score\"] = term_info[\"Similarity Score\"].values[0]\n",
    "            node_props[\"term_size\"] = int(term_info[\"Num Genes\"].values[0])\n",
    "    else:\n",
    "        node_props[\"term_name\"] = str(node_name)\n",
    "    \n",
    "    if node_name in assigned_genes:\n",
    "        node_props[\"assigned_genes\"] = assigned_genes[node_name]\n",
    "        if \"term_size\" not in node_props:\n",
    "            node_props[\"term_size\"] = len(assigned_genes[node_name])\n",
    "            node_props[\"similarity_score\"] = 0\n",
    "    \n",
    "    # This is a gene entry:\n",
    "    if node_name in gene_details:\n",
    "        details = gene_details[node_name]\n",
    "        node_props[\"sgd\"] = str(details[\"sgd\"])\n",
    "        node_props[\"full_name\"] = details[\"name\"]\n",
    "        node_props[\"synonyms\"] = details[\"syn_list\"] \n",
    "        \n",
    "    node = index.get_or_create(\"name\", node_name.lower(), node_props)\n",
    "     \n",
    "    if node_name in assigned_genes:\n",
    "        for gene in node_props[\"assigned_genes\"]:\n",
    "            index.add(\"assigned_genes\", gene.lower(), node)\n",
    "        \n",
    "    if node_name in gene_details:\n",
    "        index.add_if_none(\"sgd\", node_props[\"sgd\"].lower(), node)\n",
    "        index.add_if_none(\"full_name\", node_props[\"full_name\"].lower(), node)\n",
    "        for gene in node_props[\"synonyms\"]:\n",
    "            index.add(\"synonyms\", gene.lower(), node)\n",
    "    \n",
    "    # Extra index\n",
    "    if \"term_name\" in node_props:\n",
    "        index.add_if_none(\"term_name\", node_props[\"term_name\"].lower(), node)\n",
    "    else:\n",
    "        print(\"ERR!!!!!!!!!!!!\")\n",
    "        print(node_props)\n",
    "\n",
    "    \n",
    "    return node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Create actual DAG in database\n",
    "\n",
    "nodes = {}\n",
    "\n",
    "for row in df_graph.itertuples():\n",
    "    source_name = str(row[2])\n",
    "    target_name = str(row[1])\n",
    "    \n",
    "    if source_name in nodes:\n",
    "        source = nodes[source_name]\n",
    "    else:\n",
    "        source = add_node(source_name, assigned_genes, idx, df_term_props)\n",
    "        nodes[source_name] = source\n",
    "    \n",
    "    if target_name in nodes:\n",
    "        target = nodes[target_name]\n",
    "    else:\n",
    "        target = add_node(target_name, assigned_genes, idx, df_term_props)\n",
    "        nodes[target_name] = target\n",
    "\n",
    "    edge = Relationship(source, \"child_of\", target)\n",
    "    edge.properties[\"type\"] = \"term_relation\" \n",
    "    g.create(edge)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1521\n",
      "<generator object Index.query at 0x116a11db0>\n",
      "(n571 {full_name:\"Putative glycoside hydrolase of the mitochondrial intermembrane space\",name:\"YBR056W\",sgd:\"S000000260\",synonyms:[\"YBR056W\",\"17-beta-hydroxysteroid dehydrogenase-like protein\"],term_name:\"YBR056W\"})\n",
      "(n665 {full_name:\"Protein of unknown function\",name:\"YBR287W\",sgd:\"S000000491\",synonyms:[\"YBR287W\",\"ZSP1\"],term_name:\"YBR287W\"})\n",
      "(n698 {full_name:\"Protein with a role in ER delivery of tail-anchored membrane proteins\",name:\"YBR137W\",sgd:\"S000000341\",synonyms:[\"YBR137W\"],term_name:\"YBR137W\"})\n",
      "(n115 {full_name:\"Middle sporulation-specific mitogen-activated protein kinase (MAPK)\",name:\"SMK1\",sgd:\"S000006258\",synonyms:[\"YPR054W\",\"mitogen-activated protein kinase SMK1\"],term_name:\"SMK1\"})\n",
      "(n485 {full_name:\"Mitogen-activated protein kinase (MAPK)\",name:\"KSS1\",sgd:\"S000003272\",synonyms:[\"YGR040W\",\"mitogen-activated serine/threonine-protein kinase KSS1\"],term_name:\"KSS1\"})\n",
      "(n586 {full_name:\"MAPKKK acting in the protein kinase C signaling pathway\",name:\"BCK1\",sgd:\"S000003631\",synonyms:[\"YJL095W\",\"LAS3\",\"SAP3\",\"SLK1\",\"SSP31\",\"mitogen-activated protein kinase kinase kinase BCK1\"],term_name:\"BCK1\"})\n",
      "(n587 {full_name:\"MAPKK involved in the protein kinase C signaling pathway\",name:\"MKK1\",sgd:\"S000005757\",synonyms:[\"YOR231W\",\"SSP32\",\"mitogen-activated protein kinase kinase MKK1\"],term_name:\"MKK1\"})\n"
     ]
    }
   ],
   "source": [
    "print(g.size)\n",
    "\n",
    "# test result\n",
    "results0 = idx.query(\"name:YBR056W\")\n",
    "results1 = idx.query(\"name:ybr*\")\n",
    "\n",
    "results2 = idx.query(\"sgd:S000006124\")\n",
    "results3 = idx.query(\"synonyms:PKA2\")\n",
    "results4 = idx.query(\"full_name:*mapk*\")\n",
    "\n",
    "\n",
    "print(results1)\n",
    "for n in results1:\n",
    "    print(n)\n",
    "\n",
    "res2 = idx.query(\"term_name:'autophagy and related processes'\")\n",
    "\n",
    "for n in results4:\n",
    "    print(n)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Insert raw interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(122760, 22)\n",
      "Index(['INT1', 'INT2', 'Co-expression', 'Genetic interactions',\n",
      "       'Protein-protein interactions (high-throughput)',\n",
      "       'Domain co-occurrence', 'Genomic context', 'Phylogenetic similarity',\n",
      "       'Predicted from 3D structure', 'Conditional genetic interactions',\n",
      "       'Co-citation', 'Protein-protein interactions (low-throughput)',\n",
      "       'Co-expression.1', 'Genetic interactions.1',\n",
      "       'Protein-protein interactions (high-throughput).1',\n",
      "       'Domain co-occurrence.1', 'Genomic context.1',\n",
      "       'Phylogenetic similarity.1', 'Predicted from 3D structure.1',\n",
      "       'Conditional genetic interactions.1', 'Co-citation.1',\n",
      "       'Protein-protein interactions (low-throughput).1'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kono/anaconda/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2723: DtypeWarning: Columns (12) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INT1</th>\n",
       "      <th>INT2</th>\n",
       "      <th>Co-expression</th>\n",
       "      <th>Genetic interactions</th>\n",
       "      <th>Protein-protein interactions (high-throughput)</th>\n",
       "      <th>Domain co-occurrence</th>\n",
       "      <th>Genomic context</th>\n",
       "      <th>Phylogenetic similarity</th>\n",
       "      <th>Predicted from 3D structure</th>\n",
       "      <th>Conditional genetic interactions</th>\n",
       "      <th>...</th>\n",
       "      <th>Co-expression.1</th>\n",
       "      <th>Genetic interactions.1</th>\n",
       "      <th>Protein-protein interactions (high-throughput).1</th>\n",
       "      <th>Domain co-occurrence.1</th>\n",
       "      <th>Genomic context.1</th>\n",
       "      <th>Phylogenetic similarity.1</th>\n",
       "      <th>Predicted from 3D structure.1</th>\n",
       "      <th>Conditional genetic interactions.1</th>\n",
       "      <th>Co-citation.1</th>\n",
       "      <th>Protein-protein interactions (low-throughput).1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YBR056W</td>\n",
       "      <td>GPM2</td>\n",
       "      <td>0.798</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>22745270</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YBR056W</td>\n",
       "      <td>KKQ8</td>\n",
       "      <td>0.517</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>22289124</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YBR056W</td>\n",
       "      <td>RTK1</td>\n",
       "      <td>0.824</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>21348598</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YBR056W</td>\n",
       "      <td>YBR287W</td>\n",
       "      <td>0.483</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>24170935</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YBR056W</td>\n",
       "      <td>YET2</td>\n",
       "      <td>0.763</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>24170935</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>YBR056W</td>\n",
       "      <td>YIL055C</td>\n",
       "      <td>0.849</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>26150418</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>YBR056W</td>\n",
       "      <td>YDL027C</td>\n",
       "      <td>0.580</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>GSE30054</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>YBR056W</td>\n",
       "      <td>YDL057W</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>18854817</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>YBR056W</td>\n",
       "      <td>UGX2</td>\n",
       "      <td>0.986</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>22289124</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>YBR056W</td>\n",
       "      <td>YGR130C</td>\n",
       "      <td>0.788</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>20581845</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      INT1     INT2  Co-expression  Genetic interactions  \\\n",
       "0  YBR056W     GPM2          0.798                   0.0   \n",
       "1  YBR056W     KKQ8          0.517                   0.0   \n",
       "2  YBR056W     RTK1          0.824                   0.0   \n",
       "3  YBR056W  YBR287W          0.483                   0.0   \n",
       "4  YBR056W     YET2          0.763                   0.0   \n",
       "5  YBR056W  YIL055C          0.849                   0.0   \n",
       "6  YBR056W  YDL027C          0.580                   0.0   \n",
       "7  YBR056W  YDL057W          0.275                   0.0   \n",
       "8  YBR056W     UGX2          0.986                   0.0   \n",
       "9  YBR056W  YGR130C          0.788                   0.0   \n",
       "\n",
       "   Protein-protein interactions (high-throughput)  Domain co-occurrence  \\\n",
       "0                                             0.0                   0.0   \n",
       "1                                             0.0                   0.0   \n",
       "2                                             0.0                   0.0   \n",
       "3                                             0.0                   0.0   \n",
       "4                                             0.0                   0.0   \n",
       "5                                             0.0                   0.0   \n",
       "6                                             0.0                   0.0   \n",
       "7                                             0.0                   0.0   \n",
       "8                                             0.0                   0.0   \n",
       "9                                             0.0                   0.0   \n",
       "\n",
       "   Genomic context  Phylogenetic similarity  Predicted from 3D structure  \\\n",
       "0              0.0                      0.0                          0.0   \n",
       "1              0.0                      0.0                          0.0   \n",
       "2              0.0                      0.0                          0.0   \n",
       "3              0.0                      0.0                          0.0   \n",
       "4              0.0                      0.0                          0.0   \n",
       "5              0.0                      0.0                          0.0   \n",
       "6              0.0                      0.0                          0.0   \n",
       "7              0.0                      0.0                          0.0   \n",
       "8              0.0                      0.0                          0.0   \n",
       "9              0.0                      0.0                          0.0   \n",
       "\n",
       "   Conditional genetic interactions  \\\n",
       "0                               0.0   \n",
       "1                               0.0   \n",
       "2                               0.0   \n",
       "3                               0.0   \n",
       "4                               0.0   \n",
       "5                               0.0   \n",
       "6                               0.0   \n",
       "7                               0.0   \n",
       "8                               0.0   \n",
       "9                               0.0   \n",
       "\n",
       "                        ...                        Co-expression.1  \\\n",
       "0                       ...                               22745270   \n",
       "1                       ...                               22289124   \n",
       "2                       ...                               21348598   \n",
       "3                       ...                               24170935   \n",
       "4                       ...                               24170935   \n",
       "5                       ...                               26150418   \n",
       "6                       ...                               GSE30054   \n",
       "7                       ...                               18854817   \n",
       "8                       ...                               22289124   \n",
       "9                       ...                               20581845   \n",
       "\n",
       "   Genetic interactions.1 Protein-protein interactions (high-throughput).1  \\\n",
       "0                       0                                                0   \n",
       "1                       0                                                0   \n",
       "2                       0                                                0   \n",
       "3                       0                                                0   \n",
       "4                       0                                                0   \n",
       "5                       0                                                0   \n",
       "6                       0                                                0   \n",
       "7                       0                                                0   \n",
       "8                       0                                                0   \n",
       "9                       0                                                0   \n",
       "\n",
       "   Domain co-occurrence.1  Genomic context.1  Phylogenetic similarity.1  \\\n",
       "0                       0                  0                          0   \n",
       "1                       0                  0                          0   \n",
       "2                       0                  0                          0   \n",
       "3                       0                  0                          0   \n",
       "4                       0                  0                          0   \n",
       "5                       0                  0                          0   \n",
       "6                       0                  0                          0   \n",
       "7                       0                  0                          0   \n",
       "8                       0                  0                          0   \n",
       "9                       0                  0                          0   \n",
       "\n",
       "   Predicted from 3D structure.1  Conditional genetic interactions.1  \\\n",
       "0                              0                                   0   \n",
       "1                              0                                   0   \n",
       "2                              0                                   0   \n",
       "3                              0                                   0   \n",
       "4                              0                                   0   \n",
       "5                              0                                   0   \n",
       "6                              0                                   0   \n",
       "7                              0                                   0   \n",
       "8                              0                                   0   \n",
       "9                              0                                   0   \n",
       "\n",
       "   Co-citation.1 Protein-protein interactions (low-throughput).1  \n",
       "0              0                                               0  \n",
       "1              0                                               0  \n",
       "2              0                                               0  \n",
       "3              0                                               0  \n",
       "4              0                                               0  \n",
       "5              0                                               0  \n",
       "6              0                                               0  \n",
       "7              0                                               0  \n",
       "8              0                                               0  \n",
       "9              0                                               0  \n",
       "\n",
       "[10 rows x 22 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# New Raw interaction data \n",
    "#df_raw2 = pd.read_csv('./raw_interactions_2_w_pm_ids.txt', sep=\"\\t\");\n",
    "df_raw2 = pd.read_csv('./data/8_22_2016/intTable_kei_pm_ranks.txt', sep='\\t')\n",
    "\n",
    "print(df_raw2.shape)\n",
    "print(df_raw2.columns)\n",
    "df_raw2.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "old_df_raw2 = pd.read_csv('./raw_interactions_2_w_pm_ids.txt', sep=\"\\t\");\n",
    "print(old_df_raw2.shape)\n",
    "print(old_df_raw2.columns)\n",
    "old_df_raw2.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Co-expression', 'Genetic interactions',\n",
      "       'Protein-protein interactions (high-throughput)',\n",
      "       'Domain co-occurrence', 'Genomic context', 'Phylogenetic similarity',\n",
      "       'Predicted from 3D structure', 'Conditional genetic interactions',\n",
      "       'Co-citation', 'Protein-protein interactions (low-throughput)'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "interaction_types = df_raw2.columns[2:12]\n",
    "print(interaction_types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def create_edges(source, target, row, headers, g):\n",
    "    edges = []\n",
    "    \n",
    "    scores = row[3:13]\n",
    "    pubs = row[13:]\n",
    "    \n",
    "    for i, score in enumerate(scores):\n",
    "        if score == 0:\n",
    "            continue\n",
    "        else:\n",
    "            e = Relationship(source, headers[i], target)\n",
    "            e.properties[\"score\"] = score\n",
    "            e.properties[\"publication\"] = str(pubs[i])\n",
    "\n",
    "            e.properties[\"type\"] = \"raw_interaction\" \n",
    "            g.create(e)\n",
    "            \n",
    "def create_node(key, idx, gene_details):\n",
    "    node_dict = None\n",
    "    \n",
    "    if key in gene_details:\n",
    "        details = gene_details[key]\n",
    "        node_dict = {\n",
    "                    \"name\": key,\n",
    "                    \"term_name\": key,\n",
    "                    \"sgd\": details['sgd'],\n",
    "                    \"full_name\": details[\"name\"],\n",
    "                    \"synonyms\": details[\"syn_list\"] }\n",
    "    else:\n",
    "        node_dict = {\n",
    "           \"name\": key,\n",
    "            \"term_name\": key\n",
    "        }\n",
    "    \n",
    "    node = idx.get_or_create(\"name\", key.lower(), node_dict)\n",
    "    \n",
    "    return node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OK\n"
     ]
    }
   ],
   "source": [
    "if \"SNX4\" in gene_details:\n",
    "    print(\"OK\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for row in df_raw2.itertuples():\n",
    "    s = str(row[1])\n",
    "    t = str(row[2])\n",
    "    s_node = idx.get(\"name\", s)\n",
    "    t_node = idx.get(\"name\", t)\n",
    "    \n",
    "    if len(t_node) == 0:\n",
    "        target = create_node(t, idx, gene_details)\n",
    "    else:\n",
    "        target = t_node[0]\n",
    "        \n",
    "    if len(s_node) == 0:\n",
    "        source = create_node(s, idx, gene_details)\n",
    "    else:\n",
    "        source = s_node[0]\n",
    "    \n",
    "    create_edges(source, target, row, interaction_types, g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44555"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g.size"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
